example/mvu/README.md

```diff
@@ -0,0 +1,63 @@
+# MVU (Model-View-Update)
+
+This sample makes use of the [dartea](https://pub.dartlang.org/packages/dartea) and [built_value](https://pub.dartlang.org/packages/built_value) libraries to manage state.
+
+
+[dartea](https://pub.dartlang.org/packages/dartea) is implementation of enginge for MVU pattern inspired by [TEA (The Elm Architecture)](https://guide.elm-lang.org/architecture/) and [Elmish (F# TEA implemetation)](https://fable-elmish.github.io/elmish/)
+
+![Simple MVU app](mvu_todo.png "Simple MVU app")
+
+## Components
+* #### Model
+    This is a snapshot of your application's state, defined as an immutable data structure using the [built_value](https://pub.dartlang.org/packages/built_value) library.
+* #### Update
+    This is a pure function that produces a new state of your application given the previous state and, optionally, new commands to process.
+    ```dart
+    Upd<TModel, TMsg> update(TMsg msg, TModel model)
+    ```
+    `Upd` is simple data-container, it contains properties `model → TModel` and `effects → Cmd<TMsg>` (side-effects)
+* #### View
+    This is a pure function that produces a new UI (Widget's tree) given the current state.
+    ```dart
+    Widget view<TModel, TMsg>(BuildContext ctx, Dispatch<TMsg> dispatch, TModel model)
+    ```
+    Dispatch is simple function `void dispatch(TMsg msg)` which is dispatched `messages` into program loop.
+* #### Message
+    This an event representing a change (delta) in the state of your application, originally defined as [Union type in Elm](https://guide.elm-lang.org/types/union_types.html) or [Discriminated union in F#](https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/discriminated-unions). Dart doesn't have built-in support for [Algebraic data types](https://en.wikipedia.org/wiki/Algebraic_data_type), so we can use simple inheritance instead.
+* #### Command
+    This is a carrier of instructions, that when evaluated may produce one or more messages. It's used for controlled side-effects.
+* #### Subscriptions
+    The events from the outside world, which are translated into messages. For example data from Websockets, timers and etc.
+
+## Key concepts
+The heart of the MVU application are three yellow boxes on the diagram. First, the state of the app (Model) is mapped to the widgets tree (View). Second, events from the UI are translated into Messages and goes to the Update function (together with current app state). Update function is the brain of the app. It contains all the presentation logic, and it MUST be [pure](https://en.wikipedia.org/wiki/Pure_function). All the side-effects (such as database queries, http requests and etc) must be isolated using Commands and Subscriptions.
+
+### Pros
+* #### Easy to understand and to maintain
+    All the app state is in one immutable object (Model), all the app logic is in one pure function (Update), all the view rendering logic is in one pure function (View)
+* #### Super easy to test
```

Comment:
Could I please ask a bit more about testing and making sure I'm understanding the pattern correctly? If I follow this correctly, the update function returns an `Update`, which is a tuple of a model and effects.

However, when I looked at the tests, it seems like only half of the tuple is tested: The model. I couldn't spot any tests of the effects themselves. Are there strategies to test the Commands produced by the update function?

For example, in the update function of `/stats/state.dart`, the `LoadStats` branch of logic creates a `loadCmd`, which needs to call the repo and then dispatch another `OnStatsLoaded` Message. It feels like that logic is ripe for testing, first to make sure this command actually loads todos, and second to make sure you're doing right thing when the todos are loaded. 

What are your thoughts?

Reply:
Good question!
Yes, you are absolutely right - there are no tests for commands. I decided to exclude it from unit-tests, because I thought that integration tests should cover all the cases with commands. And, actually they covered it, but now I think that it's not right.

One way how to test commands is:
1. Create and provide commands factory to `update` function.
2. Implement this factory for tests.
3. Run command in test and verify that it triggers desired side-effect and generates correct `message`.

I'm going to try this approach today, seems like it should work.